runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/budget/GetWorkBudget.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.windmill.work.budget;
+
+import com.google.auto.value.AutoValue;
+import org.apache.beam.runners.dataflow.worker.windmill.Windmill.GetWorkRequest;
+import org.apache.beam.runners.dataflow.worker.windmill.streams.WindmillStream;
+
+/**
+ * Budget of items and bytes for fetching {@link
+ * org.apache.beam.runners.dataflow.worker.windmill.Windmill.WorkItem}(s) via {@link
+ * WindmillStream.GetWorkStream}. Used to control how "much" work is returned from Windmill.
+ */
+@AutoValue
+public abstract class GetWorkBudget {
+  public static GetWorkBudget.Builder builder() {
+    return new AutoValue_GetWorkBudget.Builder();
+  }
+
+  /** {@link GetWorkBudget} of 0. */
+  public static GetWorkBudget noBudget() {
+    return builder().setItems(0).setBytes(0).build();
+  }
+
+  public static GetWorkBudget from(GetWorkRequest getWorkRequest) {
+    return builder()
+        .setItems(getWorkRequest.getMaxItems())
+        .setBytes(getWorkRequest.getMaxBytes())
+        .build();
+  }
+
+  /**
+   * Adds the given bytes and items or the current budget, returning a new {@link GetWorkBudget}.
+   * Does not drop below 0.
+   */
+  public GetWorkBudget add(long items, long bytes) {
+    return GetWorkBudget.builder()
+        .setBytes(Math.max(0, bytes() + bytes))
```

Comment:
build takes care of max, remove here and in subtract?

Reply:
done